flow/aggregate.py

```diff
@@ -0,0 +1,183 @@
+        if aggregator is None:
+            def aggregator(jobs):
+                return jobs
+
+        if not callable(aggregator):
+            raise TypeError("Expected callable aggregator function, got {}"
+                            "".format(type(aggregator)))
+
+        if sort is not None and not isinstance(sort, str):
+            raise TypeError("Expected string sort parameter, got {}".format(type(sort)))
+
+        if not isinstance(reverse, bool):
+            raise TypeError("Expected bool reverse parameter got {}".format(type(reverse)))
+
+        if select is not None and not callable(select):
+            raise TypeError("Expected callable select parameter, got {}".format(type(select)))
+
+        self._aggregator = aggregator
+        self._sort = None if sort is None else partial(sorted,
+                                                       key=lambda job: job.sp[sort],
+                                                       reverse=reverse)
+        self._select = None if select is None else partial(filter, select)
+
+    @classmethod
+    def groupsof(cls, num=1, sort=None, reverse=False, select=None):
+        # copied from: https://docs.python.org/3/library/itertools.html#itertools.zip_longest
+        try:
+            num = int(num)
+            if num < 0:
+                raise ValueError("The num parameter should be greater than 0")
+        except Exception:
+            raise TypeError("The num parameter should be an integer")
+
+        def aggregator(jobs):
+            args = [iter(jobs)] * num
+            return zip_longest(*args)
+
+        return cls(aggregator, sort, reverse, select)
+
+    @classmethod
+    def groupby(cls, key, default=None, sort=None, reverse=False, select=None):
+        if isinstance(key, str):
+            if default is None:
+                def keyfunction(job):
+                    return job.sp[key]
+            else:
+                def keyfunction(job):
+                    return job.sp.get(key, default)
+
+        elif isinstance(key, Iterable):
+            keys = list(key)
+
+            if default is None:
+                def keyfunction(job):
+                    return [job.sp[key] for key in keys]
+            else:
+                def keyfunction(job):
+                    return [job.sp.get(key, default) for key in keys]
+
+        elif callable(key):
+            keyfunction = key
+
+        else:
+            raise TypeError("Invalid key argument. Expected either str, Iterable "
+                            "or a callable, got {}".format(type(key)))
+
+        def aggregator(jobs):
+            for key, group in groupby(sorted(jobs, key=keyfunction), key=keyfunction):
+                yield group
+
+        return cls(aggregator, sort, reverse, select)
+
+    def __call__(self, obj=None):
+        if callable(obj):
+            setattr(obj, '_flow_aggregate', self)
+            return obj
+        elif isinstance(obj, (list, Project, JobsCursor)):
```

Comment:
This conditional makes me thing we should be dealing with two classes, one for registering the aggregates and another for performing the logic associated with them.